src/util/dapps.js

```diff
@@ -0,0 +1,245 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import BigNumber from 'bignumber.js';
+import { pick, range, uniq } from 'lodash';
+
+import { bytesToHex } from '@parity/api/lib/util/format';
+import { IconCache } from '@parity/ui';
+import isElectron from 'is-electron';
+
+import builtinApps from '../Dapps/dappsBuiltin.json';
+import Contracts from '@parity/shared/lib/contracts';
+
+import path from 'path';
+
+const util = isElectron() ? window.require('util') : require('util');
+
+require('util.promisify').shim();
+
+const fs = isElectron() ? window.require('fs') : require('fs');
+const fsReadFileAsync = util.promisify(fs.readFile);
+
+export function subscribeToChanges (api, dappReg, callback) {
+  return dappReg
+    .getContract()
+    .then((dappRegContract) => {
+      const dappRegInstance = dappRegContract.instance;
+
+      const signatures = ['MetaChanged', 'OwnerChanged', 'Registered']
+        .map((event) => dappRegInstance[event].signature);
+
+      return api.eth
+        .newFilter({
+          fromBlock: '0',
+          toBlock: 'latest',
+          address: dappRegInstance.address,
+          topics: [signatures]
+        })
+        .then((filterId) => {
+          return api
+            .subscribe('eth_blockNumber', () => {
+              if (filterId > -1) {
+                api.eth
+                  .getFilterChanges(filterId)
+                  .then((logs) => {
+                    return dappRegContract.parseEventLogs(logs);
+                  })
+                  .then((events) => {
+                    if (events.length === 0) {
+                      return [];
+                    }
+
+                    // Return uniq IDs which changed meta-data
+                    const ids = uniq(events.map((event) => bytesToHex(event.params.id.value)));
+
+                    callback(ids);
+                  });
+              }
+            })
+            .then((blockSubId) => {
+              return {
+                block: blockSubId,
+                filter: filterId
+              };
+            });
+        });
+    });
+}
+
+export function fetchBuiltinApps (api) {
+  const basePath = isElectron() ? window.require('electron').remote.getGlobal('dirName') : path.join(__dirname, '..');
+  const initialApps = builtinApps.filter(app => app.id);
+
+  // Replace all backslashes by front-slashes (happens in Windows)
+  // Note: `dirName` contains backslashes in Windows. One would assume that
+  // path.join in Windows would handle everything for us, but after some time
+  // I realized that even in Windows path.join here bahaves like POSIX (maybe
+  // it's electron, maybe browser env?). Switching to '/'. -Amaury 12.03.2018
+  const posixDirName = basePath.replace(/\\/g, '/');
```

Comment:
This piece of code is now in 2 places. We could add another file inside `util/` to factorize that.